src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt

```diff
@@ -1220,30 +1221,34 @@ private fun processLexicalDeclarations(
     return false
 }
 
+fun processNestedScopesUpwards(scopeStart: RsElement, ns: Set<Namespace>, processor: RsResolveProcessor): Boolean =
+    processNestedScopesUpwards(scopeStart, ns, isCompletion = false, processor = processor)
+
 fun processNestedScopesUpwards(
     scopeStart: RsElement,
     ns: Set<Namespace>,
+    isCompletion: Boolean,
     processor: RsResolveProcessor
 ): Boolean {
     val prevScope = mutableSetOf<String>()
     if (walkUp(scopeStart, { it is RsMod }) { cameFrom, scope ->
-        val currScope = mutableListOf<String>()
-        val shadowingProcessor = { e: ScopeEntry ->
-            e.name !in prevScope && run {
-                currScope += e.name
-                processor(e)
-            }
+        processWithShadowing(prevScope, processor) { shadowingProcessor ->
+            processLexicalDeclarations(scope, cameFrom, ns, shadowingProcessor)
         }
-        if (processLexicalDeclarations(scope, cameFrom, ns, shadowingProcessor)) return@walkUp true
-        prevScope.addAll(currScope)
-        false
     }) {
         return true
     }
 
     // Prelude is injected via implicit star import `use std::prelude::v1::*;`
     if (processor(ScopeEvent.STAR_IMPORTS)) return false
 
+    if (Namespace.Types in ns) {
+        val result = processWithShadowing(prevScope, processor) { shadowingProcessor ->
+            processExternCrateResolveVariants(scopeStart, isCompletion, shadowingProcessor)
```

Comment:
Hm, interesting
The following code
```rust
fn main() {
    extern_crate_name::some_function_name();
}
```
really works with 2015 edition (with the current stable compiler) even without `extern crate` item. When did it become possible?

Reply:
With Rust 1.30, see https://blog.rust-lang.org/2018/10/25/Rust-1.30.0.html#module-system-improvements

(and we support it already, I just moved the code from another place to here)